### PR TITLE
Creating test_Pipelines.py

### PR DESCRIPTION
**Pull request by niravkin**
This pull request is creating test_Pipelines.py, which tests all functions in Pipelines.py.
There were no special assumptions made about any function, and I tested all functions for normal cases and tried to find invalid inputs that would raise exceptions.

Added import for the mock library in python-package.yml to allow the use of mock with pytest.

### CODE REVIEW

File: pyCGM_Single/tests/test_Pipelines.py [hunk 1]
```diff
@@ -0,0 +1,423 @@
...
+            result = Pipelines.butterFilter(data, cutoff, Fs)
+            numpyResult = Pipelines.butterFilter(numpyData, cutoff, Fs)
+            expectedResult = expectedResults[i]
+            for i in range(len(result)):
+                np.testing.assert_almost_equal(result[i], expectedResult[i], self.rounding_precision)
+                np.testing.assert_almost_equal(numpyResult[i], expectedResult[i], self.rounding_precision)
+
+        #Test that if len(data) < 15, an exception is raised
+        dataLenTests = [[],
+                        [0, 1, 2, 3],
```
Comment (cadop) [inline pyCGM_Single/tests/test_Pipelines.py:65]: this is unnecessarily verbose.  You could simply use range and np.linspace

File: pyCGM_Single/tests/test_Pipelines.py [hunk 2]
```diff
@@ -0,0 +1,423 @@
...
+            for i in range(len(result)):
+                np.testing.assert_almost_equal(result[i], expectedResult[i], self.rounding_precision)
+                np.testing.assert_almost_equal(numpyResult[i], expectedResult[i], self.rounding_precision)
+
+        #Test that if len(data) < 15, an exception is raised
+        dataLenTests = [[],
+                        [0, 1, 2, 3], 
+                        [0.1, 0.2, 0.3], 
+                        [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14]]
+        for test in dataLenTests:
```
Comment (cadop) [inline pyCGM_Single/tests/test_Pipelines.py:68]: Please test one thing at a time

**Review by cadop — CHANGES_REQUESTED** [review]

File: pyCGM_Single/tests/test_Pipelines.py (hunk 2 above)
Comment (niravkin) [reply pyCGM_Single/tests/test_Pipelines.py:68]: Could you clarify which things I am testing at the same time? I wrote this to test four different instances where the length of data is too short and made sure that an exception was raised.

File: pyCGM_Single/tests/test_Pipelines.py [hunk 3]
```diff
@@ -0,0 +1,517 @@
...
+
+We test cases where inputs are positive and negative floats and 
+integers. We test cases where inputs are lists and numpy arrays.
+
+We test exceptions raised when the length of the input data is too short, 
+when the cutoff frequency value is negative, zero, or too large, 
+and when the sampling frequency value is negative, zero, or too small.
+"""
+#Test positive and negative floats and ints
+listTests = [
```
Comment (cadop) [inline pyCGM_Single/tests/test_Pipelines.py:49]: Please keep pep8 naming (except when its not possible because of pycgm)

File: pyCGM_Single/tests/test_Pipelines.py [hunk 4]
```diff
@@ -0,0 +1,517 @@
...
+    (range(20), 20, -10),
+    (range(20), 20, 0),
+    (range(20), 20, 20),
+    (range(20), 20, 60)
+])
+def test_butterFilterExceptions(data, cutoff, Fs):
+    with pytest.raises(Exception):
+        Pipelines.butterFilter(data, cutoff, Fs)
+
+"""
```
Comment (cadop) [inline pyCGM_Single/tests/test_Pipelines.py:111]: What is this string for? It seems like a docstring, but is below one function and above another.  It may be clearer to use it as a docstring.

File: pyCGM_Single/tests/test_Pipelines.py [hunk 5]
```diff
@@ -0,0 +1,517 @@
...
+
+We test the exception raised when the marker SACR does not
+exist in Data.
+"""
+data,static = load_data_from_file(3, calcSacrum=True) #True indicates we calculate SACR before testing
+framesToClear = [1, 10, 12, 15, 100, -1]
+for frameNum in framesToClear:
+    data['LFHD'][frameNum] = np.array([np.nan, np.nan, np.nan])
+
+result = Pipelines.rigid_fill(data, static)
```
Comment (cadop) [inline pyCGM_Single/tests/test_Pipelines.py:498]: this should be in a function.

**Review by cadop — CHANGES_REQUESTED** [review]
I think it would help to follow the pytest guides on making a class, which would solve many of the strange list comprehensions and global variables going on.

File: pyCGM_Single/tests/test_Pipelines.py (hunk 3 above)
Comment (niravkin) [reply pyCGM_Single/tests/test_Pipelines.py]: Changed variables to follow pep8 naming.

File: pyCGM_Single/tests/test_Pipelines.py (hunk 4 above)
Comment (niravkin) [reply pyCGM_Single/tests/test_Pipelines.py]: Changed the string to appear inside of the testing functions so it is clearer as a docstring.

File: pyCGM_Single/tests/test_Pipelines.py (hunk 5 above)
Comment (niravkin) [reply pyCGM_Single/tests/test_Pipelines.py]: Changed rigid_fill to be called inside of the test_rigid_fill_accuracy function.

File: pyCGM_Single/tests/test_Pipelines.py [hunk 6]
```diff
@@ -0,0 +1,534 @@
...
+        ({'trajOne': [[217.19961548, -82.35484314, 332.2684021 ],
+                    [257.19961548, -32.35484314, 382.2684021 ]]})
+    ])
+    def test_prep_exceptions(self, trajs):
+        with pytest.raises(Exception):
+            Pipelines.prep(trajs)
+    
+    @pytest.mark.parametrize("data, name, expected_result", [
+        (
+        [{'LTIL': np.array([-268.1545105, 327.53512573,  30.17036057]),
```
Comment (cadop) [inline pyCGM_Single/tests/test_Pipelines.py:230]: please refine the formatting here a bit, its hard to follow

**Review by cadop — CHANGES_REQUESTED** [review]
Overall well done. Please just fix the one section for formatting to make it easier to follow/read

File: pyCGM_Single/tests/test_Pipelines.py [hunk 7]
```diff
@@ -0,0 +1,627 @@
...
+        We test the exception that is raised when the marker name does not
+        exist.
+        """
+        data = self.data
+        with pytest.raises(Exception):
+            Pipelines.segmentFinder(key,data,target_dict,segment_dict,j,missings)
+    
+    def test_rigid_fill_transform_from_static(self):
+        """
+        This function tests Pipelines.rigid_fill(Data, static),
```
Comment (cadop) [inline pyCGM_Single/tests/test_Pipelines.py:558]: There are 4 separate paragraphs saying 'this function' does something different

**Review by cadop — CHANGES_REQUESTED** [review]

**Review by cadop — APPROVED** [review]
